lib/helpers/createServerMiddleware.js

```diff
@@ -0,0 +1,47 @@
+const admin = require('firebase-admin')
+const logger = require('../logger')
+
+module.exports = function (options) {
+  return async (req, res, next) => {
+    if (!req.headers.authorization) { return next() }
+
+    const credential = options.credential
+      ? admin.credential.cert(options.credential)
+      : admin.credential.applicationDefault()
+
+    // initialize admin sdk if not yet initialized
+    if (!admin.apps.length) {
+      admin.initializeApp({
+        credential,
+        ...options.config
+      })
+    }
+
+    // Parse the injected ID token from the request header.
+    const authorizationHeader = req.headers.authorization || ''
+    const components = authorizationHeader.split(' ')
+    const idToken = components.length > 1 ? components[1] : ''
+
+    try {
+      // Try to verify the id token, additionally checking if the token was revoked
+      const decodedToken = await admin.auth().verifyIdToken(idToken, true)
+
+      // Get user record
+      const userRecord = await admin.auth().getUser(decodedToken.uid)
+
+      // separate properties preserve established API
+      const { customClaims, ...authUser } = userRecord
+
+      // @TODO: deprecate either response or request object augmentation on next major version bump
```

Comment:
I reckon you prefer using `req` rather than `res` to pass on data.

I was considering `res.locals` because according to [here](https://stackoverflow.com/questions/18875292/passing-variables-to-the-next-middleware-using-next-in-express-js) this is the recommended way to do it.

Others prefer `req`. 

I assume `res.locals` would be the most official way to do it, but `req` also works I guess. Then again I'm not a NodeJs specialst.

Any arguments for either or?

Reply:
I consider authentication & authorization to be request augmentations for that reason i prefer using `req` to set resolved users on.

The way I see it, it belongs on the `req` because logically it is information concerning the issued request, not the response, i.e. user X (`verifiedFireAuthUser`) has issued the request (`req`) and should therefore be associated with the request object.

However it's your module (thanks for the great work, btw :sweat_smile: ) so the decision is yours.

Reply:
@mentAl-maZe 
Yes that kind of makes sense to me, too.

At the same time [here](http://expressjs.com/en/api.html#res.locals) the Express team uses `res.locals`, and not `req.locals` or `req` for passing exactly these kind of authentication data, no?

> _exposing request-level information such as the request path name, authenticated user, user settings, and so on._"

Hmm... I'm happy with either or. Small detail, we can also decide to just do both like it is now ;-)

> However it's your module (thanks for the great work, btw 😅 ) so the decision is yours.

Bah, it's open source, I do listen to people who know better :P

Reply:
> At the same time here the Express team uses res.locals, and not req.locals or req for passing exactly these kind of authentication data, no?

I guess I'll have to reconsider than.
I'm all for community standards and if this is how the core team of the underlying technology recommends doing it I think we should follow suit.